src/components/views/dialogs/SpotlightDialog.tsx

```diff
@@ -175,23 +178,96 @@ function refIsForRecentlyViewed(ref: RefObject<HTMLElement>): boolean {
     return ref.current?.id.startsWith("mx_SpotlightDialog_button_recentlyViewed_");
 }
 
+enum Section {
+    People,
+    Rooms,
+    Spaces,
+}
+
+interface IBaseResult {
+    section: Section;
+    query?: string[]; // extra fields to query match, stored as lowercase
+}
+
+interface IRoomResult extends IBaseResult {
+    room: Room;
+}
+
+interface IResult extends IBaseResult {
+    avatar: JSX.Element;
+    name: string;
+    description?: string;
+    onClick?(): void;
+}
+
+type Result = IRoomResult | IResult;
+
 const SpotlightDialog: React.FC<IProps> = ({ initialText = "", onFinished }) => {
     const cli = MatrixClientPeg.get();
     const rovingContext = useContext(RovingTabIndexContext);
     const [query, _setQuery] = useState(initialText);
     const [recentSearches, clearRecentSearches] = useRecentSearches();
 
-    const results = useMemo<Room[] | null>(() => {
-        if (!query) return null;
+    const possibleResults = useMemo<Result[]>(() => [
+        ...SpaceStore.instance.enabledMetaSpaces.map(spaceKey => ({
+            section: Section.Spaces,
+            avatar: (
+                <div className={`mx_SpotlightDialog_metaspaceResult mx_SpotlightDialog_metaspaceResult_${spaceKey}`} />
+            ),
+            name: getMetaSpaceName(spaceKey, SpaceStore.instance.allRoomsInHome),
+            onClick() {
+                SpaceStore.instance.setActiveSpace(spaceKey);
+            },
+        })),
+        ...cli.getVisibleRooms().map(room => {
+            let section: Section;
+            let query: string[];
+
+            const otherUserId = DMRoomMap.shared().getUserIdForRoomId(room.roomId);
+            if (otherUserId) {
+                section = Section.People;
+                query = [
+                    otherUserId.toLowerCase(),
+                    room.getMember(otherUserId)?.name.toLowerCase(),
+                ].filter(Boolean);
+            } else if (room.isSpaceRoom()) {
+                section = Section.Spaces;
+            } else {
+                section = Section.Rooms;
+            }
+
+            return { room, section, query };
+        }),
+    ], [cli]);
+
+    const trimmedQuery = query.trim();
+    const [people, rooms, spaces] = useMemo<[Result[], Result[], Result[]] | []>(() => {
+        if (!trimmedQuery) return [];
 
-        const trimmedQuery = query.trim();
         const lcQuery = trimmedQuery.toLowerCase();
         const normalizedQuery = normalize(trimmedQuery);
 
-        return cli.getVisibleRooms().filter(r => {
-            return r.getCanonicalAlias()?.includes(lcQuery) || r.normalizedName.includes(normalizedQuery);
+        const results: [Result[], Result[], Result[]] = [[], [], []];
+
+        possibleResults.forEach(entry => {
+            if ((entry as IRoomResult).room) {
```

Comment:
This could be a nice place to use a type guard and avoid casting. Something like 
```
const isRoomResult = (result: any): result is IRoomResult => !!result?.room;
```